### DISCUSSION TITLE
Should accept wait infinitely if there is no connection in pending queue?

### DISCUSSION THREAD

**Discussion by jsjclink** [ANLAB-KAIST/KENSv3#63] [Project Q&A 2. Transport Layer (TCP)] [answered]
Hi,
In #58, and backlog limits the maximum number of current handshaking.
In backlog1, since syn is delivered almost at the same time, and backlog is 3, I think I should handshake only first 3 syn.
So in pending queue there is only 3 established connection, therefore only 3 accept return 0 and other accept call should wait infinitely.

As I expected, when I ran the testcase after 3 accept, accept waited infinitely. But the problem is that I can't  reach next testcase because of infinite loop.
So, to resolve this problem, I made accept to return after few millisecond with -1. Then I could reach next testcase, but I'm not sure it is right way..

* * *
안녕하세요
#58을 보고 backlog가 동시에 진행될 수 있는 최대 handshaking의 수를 제한한다고 생각했습니다.
그래서 backlog1에서는 syn이 client에서 거의 동시에 6개가 도착하고, backlog가 3이기 때문에 처음 3개는 handshaking을 해주고
나머지는 답장을 하지 않아야 한다고 생각했습니다.
따라서 accept가 connection을 pop하는 pending queue에는 established된 connection이 3개만 들어있게 되고, 그 결과 처음 3번의 accept에서는 문제없이 넘어가고
나머지 accept에서는 무한루프가 걸려야 할 것이라고 생각했습니다.

테스트케이스를 돌렸을 때 예상한 대로 무한루프가 걸렸습니다. 그런데 문제는 무한루프로 인해 그 뒤로 넘어가지 않는다는 것입니다.
따라서 테스트케이스를 통과하기 위해서는 일정 시간이 지난 후에도 pending queue가 비어있으면 무한루프를 빠져나오게 해야 할 것 같은데, 이것이 accept의 정상적인 동작 방식인지 모르겠습니다..

**Comment by Pusnow** [comment, marked as answer]
In `finalize()`, you should return all blocked system calls.

https://github.com/ANLAB-KAIST/KENSv3/blob/c95d8f9a7669a491317d53eb4ff2db96ad8a0ffc/app/kens/TCPAssignment.cpp#L27

**Reply by jsjclink** [reply]
thank you!
